js/src/forum/states/PostStreamState.js

```diff
@@ -0,0 +1,376 @@
+     *
+     * @type {String}
+     */
+    this.description = '';
+
+    this.show(includedPosts);
+  }
+
+  /**
+   * Load and scroll to a post with a certain number.
+   *
+   * @param {Integer|String} number The post number to go to. If 'reply', go to
+   *     the last post and scroll the reply preview into view.
+   * @param {Boolean} noAnimation
+   * @return {Promise}
+   */
+  goToNumber(number, noAnimation) {
+    console.log('goToNumber', number);
+    // If we want to go to the reply preview, then we will go to the end of the
+    // discussion and then scroll to the very bottom of the page.
+    if (number === 'reply') {
+      return this.goToLast();
+    }
+
+    this.loadPromise = this.loadNearNumber(number);
+
+    this.paused = true;
+    this.needsScroll = true;
+    this.noAnimationScroll = noAnimation;
+    this.locationType = 'number';
+    this.number = number;
+
+    // In this case, the redraw is only called after the response has been loaded
+    // because we need to know the indeces of the post range before we can
+    // start scrolling to items. Calling redraw early causes issues.
+    // Since this is only used for external navigation to the post stream, the delay
+    // before the stream is moved is not an issue.
+    return this.loadPromise.then(() => m.redraw());
+  }
+
+  /**
+   * Load and scroll to a certain index within the discussion.
+   *
+   * @param {Integer} index
+   * @param {Boolean} backwards Whether or not to load backwards from the given
+   *     index.
+   * @param {Boolean} noAnimation
+   * @return {Promise}
+   */
+  goToIndex(index, noAnimation) {
+    console.log('goToIndex', index);
+    this.paused = true;
+
+    this.loadPromise = this.loadNearIndex(index);
+
+    this.paused = true;
+    this.needsScroll = true;
+    this.noAnimationScroll = noAnimation;
+    this.locationType = 'index';
+    this.index = index;
+
+    m.redraw();
+
+    return this.loadPromise;
+  }
+
+  /**
+   * Load and scroll up to the first post in the discussion.
+   *
+   * @return {Promise}
+   */
+  goToFirst() {
+    return this.goToIndex(0);
+  }
+
+  /**
+   * Load and scroll down to the last post in the discussion.
+   *
+   * @return {Promise}
+   */
+  goToLast() {
+    return this.goToIndex(this.count() - 1);
+  }
+
+  /**
+   * Set up the stream with the given array of posts.
+   *
+   * @param {Post[]} posts
+   */
+  show(posts) {
+    this.visibleStart = posts.length ? this.discussion.postIds().indexOf(posts[0].id()) : 0;
+    this.visibleEnd = this.sanitizeIndex(this.visibleStart + posts.length);
+  }
+
+  /**
+   * Reset the stream so that a specific range of posts is displayed. If a range
+   * is not specified, the first page of posts will be displayed.
+   *
+   * @param {Integer} [start]
+   * @param {Integer} [end]
+   */
+  reset(start, end) {
+    this.visibleStart = start || 0;
+    this.visibleEnd = this.sanitizeIndex(end || this.constructor.loadCount);
```

Comment:
Since doing a major rewrite, might want to also change visibleEnd to:
```
this.visibleEnd = this.sanitizeIndex(end || this.constructor.loadCount - 1)
```

Logic:
Suppose loadCount is 1, visibleStart = 0 and visibleEnd should be the same post 0.

Would need to also adjust all the code that references visibleEnd - I can send you a pull request for that if you want.